### ISSUE TITLE
Defect Localization: ability to specify test dependencies

### ISSUE THREAD

**Issue by Fleshgrinder** [kotest/kotest#3868]
Tests often build on top of each other, and if a test fails that others rely upon to succeed we're looking at possibly thousands of failed tests. It should be obvious that it can be hard to find the root cause.

I think that [JExample](https://scg.unibe.ch/research/jexample) is the most prominent project that nails what I'm asking for. Instead of the annotations and strings it would obviously be nice to have some nifty Kotlin magic to make it ergonomic. The first thought that comes to mind is that tests simply return something, and that something can be stored in a variable. Subsequent tests can then be skipped based on the outcome, or access whatever the previous test returned. I haven't thought much about this, but maybe …

```kotlin
class MyTest : StringSpec({
    val a = "first test" {
        // do something
        1 to 2
    }

    a?."only if a succeeds" {
        // do something
        val (x, y) = a.result
        x shouldBe y
    }
})
```

**Comment by Kantis** [comment]
Does this imply that test `a` modifies state which the next test depends upon? 

Couldn't you just order the tests within the spec and enable `failFast`? (perhaps only applicable for simple cases?)

**Comment by Fleshgrinder** [comment]
There are two parts to this.

## Pure Defect Localization

We define that a test (child) depends on another test (parent). If the parent fails the child is automatically skipped with the info that it makes no sense to execute it since its parent failed.

## State Sharing

Tests that build on top of each other and share state. Either indirectly (Test Containers) or directly (the value the previous test returns).

An example I'm having here is a Kafka Streams test with Test Containers where I'm publishing messages—perform some assertions—shut down streams to then go to the next test where I verify that it restores correctly. This goes on and each test builds on each other. I tried sequential test execution, but this fails, because the tests are executed with coroutines and I'm facing random failures. My workaround is to use a single test sprinkled with `withClue("") {}` calls. But, that's not ideal, because I cannot easily run it only up to a certain point.

**Comment by sksamuel** [comment]
You could nest the tests, and add fail fast.
Then you could make the top level tests run in parallel, but tests inside each context run in sequence.

**Comment by sksamuel** [comment]
It would also be possible to have a test be defined to require another test first, like you want, but it would have to be based on strings. I don't think there's much we can do to add type safety, as things like `test("foo") { }` are not available to the type system.

I think the context grouping with fail fast per context is the way to go.

**Comment by Fleshgrinder** [comment]
Finally able to report back as I had the need for this again. The nesting of tests in contexts works as you proposed @sksamuel. The ability to return from a test was sorely missing. I always had to work with nullable `var`s and `!!` to make it clear that something is being set up by a prior test.

The fail fast was rather manual, since there's nothing to support it. [JUnit has their assumptions](https://junit.org/junit5/docs/current/user-guide/#writing-tests-assumptions), but that would be very manual as well. Since the assumption works with booleans and the reasoning must be typed out.

I still think that my initial proposal could work, and I might be inclined to play around a bit if you're open for contributions in this direction. A test could return a `TestResult` that captures some meta information about the test and optionally a return value. Context receivers could be used to bring things together, or by simply adding an optional argument to test functions.

```kotlin
class SomeTest : FunSpec({
    val a = test("a") { /*...*/ }

    // with context receivers
    a.test("b") { /*...*/ }

    // without context receivers
    test("c", dependsOn = a) { /*...*/ }
})
```

**Comment by OliverO2** [comment]
Inter-test dependencies sound like we'd risk venturing into Gradle-like magic where we ask ourselves "what gets evaluated when?"

Also, context receivers are discontinued and [will be replaced with context parameters](https://youtrack.jetbrains.com/issue/KT-67119).

**Comment by Fleshgrinder** [comment]
Why do you think so? Isn't it obvious that `a` is evaluated before `b`? Would be interesting to understand why you don't understand the dependency in the example I gave so that it can be addressed. Nobody wants weird magic. It needs to be very, very clear that one test depends on another.

Yeah, I heard about but didn't remember the new name. From a usage perspective things stay the same. All we want is that `TestResult.test` has access to the spec or container surrounding it. Which is a given, since you can only ever retrieve a `TestResult` by being inside a spec or container.

**Comment by OliverO2** [comment]
Initially, things _seem_ obvious to the naive reader, who would create a mental model in which, basically, everything executes sequentially.

However, tests are not executed directly, but merely registered. Afterwards, the test engine has full control about reordering (and even parallelizing) test executions. This is not an issue if tests are properly isolated.

Now `dependsOn` comes in. It creates an implied execution graph which is not reflected in the code structure. In simple scenarios, it's not much of an issue. I would expect this to grow into tests having multiple dependencies and transitive dependencies. As complex dependency trees come into play, things can quickly get messy. This might get burdensome to support. We already have a lightweight set of support issues with unexpected sequencing of `before...` and `after...` invocations.

We can expect the initial version of context parameters to arrive around May, 2025. Nothing to bet on right now.

**Comment by Fleshgrinder** [comment]
So it's more about the internals than users in that case. The example you gave is about inexperienced users who generally don't know how test execution works. Probably users who don't even know that they can run their tests concurrently as well as in parallel.

> […] is not reflected in the code structure.

Actually, it is - and always forced from top unlike in JUnit - but yes, a single test could accumulate a chain of dependencies within the same spec or the same container if the feature is limited to containers. JExample allows entire specs to depend on other specs. Although this makes sense with the goals they've set for themselves, it's not something I'm asking for here.

Allowing tests to return anything would already be good enough. At least then, it could be built from the outside. But, feel free to close the issue if the general stance is that it's too complex to implement in Kotest.

**Comment by Fleshgrinder** [comment]
I worked with this a little longer and the ability to return from a test function inside a container (e.g., `FunSpec.context`) paired with fail fast would really cover like 99% of the need here. Anything else would just be icing on the cake and isn't really necessary.

The ability to return from a root test makes no sense, due to the execution order issues already mentioned. Imho this is totally fine.

**Comment by sksamuel** [comment]
Closing, as not feasible to come up with a dsl that works with junit
